From d4aee32535f5f31f48b1b6e8ece30a163a24f11d Mon Sep 17 00:00:00 2001
From: Ioana Radulescu <ruxandra.radulescu@nxp.com>
Date: Wed, 13 Mar 2019 19:12:47 +0200
Subject: [PATCH 108/235] staging: fsl-dpaa2/mac: add pause frames support for
 managed phys

Read the configured dpmac options and depending on the supported
features set the according advertising bit.

Signed-off-by: Ioana Ciornei <ioana.ciornei@nxp.com>
Signed-off-by: Bogdan Purcareata <bogdan.purcareata@nxp.com>
Signed-off-by: Ioana Radulescu <ruxandra.radulescu@nxp.com>
[Xulin: Original patch taken from NXP LSDK-19.06.]
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/staging/fsl-dpaa2/mac/mac.c | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/drivers/staging/fsl-dpaa2/mac/mac.c b/drivers/staging/fsl-dpaa2/mac/mac.c
index 423c37e67c91..a03207e28605 100644
--- a/drivers/staging/fsl-dpaa2/mac/mac.c
+++ b/drivers/staging/fsl-dpaa2/mac/mac.c
@@ -154,6 +154,12 @@ static void dpaa2_mac_link_changed(struct net_device *netdev)
 		if (phydev->autoneg)
 			state.options |= DPMAC_LINK_OPT_AUTONEG;
 
+		if (phydev->pause && (phydev->advertising & ADVERTISED_Pause))
+			state.options |= DPMAC_LINK_OPT_PAUSE;
+		if (phydev->pause &&
+		    (phydev->advertising & ADVERTISED_Asym_Pause))
+			state.options |= DPMAC_LINK_OPT_ASYM_PAUSE;
+
 		netif_carrier_on(netdev);
 	} else {
 		netif_carrier_off(netdev);
@@ -416,6 +422,20 @@ static void configure_link(struct dpaa2_mac_priv *priv,
 	phydev->speed = cfg->rate;
 	phydev->duplex  = !!(cfg->options & DPMAC_LINK_OPT_HALF_DUPLEX);
 
+	if (phydev->supported & SUPPORTED_Pause) {
+		if (cfg->options & DPMAC_LINK_OPT_PAUSE)
+			phydev->advertising |= ADVERTISED_Pause;
+		else
+			phydev->advertising &= ~ADVERTISED_Pause;
+	}
+
+	if (phydev->supported & SUPPORTED_Asym_Pause) {
+		if (cfg->options & DPMAC_LINK_OPT_ASYM_PAUSE)
+			phydev->advertising |= ADVERTISED_Asym_Pause;
+		else
+			phydev->advertising &= ~ADVERTISED_Asym_Pause;
+	}
+
 	if (cfg->advertising != 0) {
 		phydev->advertising = 0;
 		link_mode_dpmac2phydev(cfg->advertising, &phydev->advertising);
-- 
2.17.1

